ext/crates/fp/src/vector/mod.rs

```diff
@@ -110,138 +110,79 @@ mod test {
         ]
     }
 
-    /// An arbitrary number of dimensions. This makes the functions that create vectors DRY.
-    fn arb_dim() -> impl Strategy<Value = usize> {
-        0usize..=10_000
-    }
-
-    /// An arbitrary pair of dimensions, where the second dimension is bigger than or equal to the
-    /// first. This happens to be the same as `arb_slice(10000)`, but we make it a separate function
-    /// because the semantics are different.
-    fn arb_dim_pair() -> impl Strategy<Value = (usize, usize)> {
-        arb_slice(10_000)
+    /// An arbitrary (prime, dimension) pair
+    fn arb_prime_dim() -> impl Strategy<Value = (ValidPrime, usize)> {
+        arb_prime().prop_flat_map(|p| (Just(p), 0usize..=10_000))
     }
 
     /// The start and end positions of an arbitrary slice of a vector of length `dimension`
     fn arb_slice(dimension: usize) -> impl Strategy<Value = (usize, usize)> {
-        // We have every integer twice because `subsequence` returns distinct values, but we want to
-        // also test the case where the slices have the same start and end positions.
-        let all_indices: Vec<_> = (0..=dimension).flat_map(|i| [i, i]).collect();
-        proptest::sample::subsequence(all_indices, 2).prop_map(|v| (v[0], v[1]))
-    }
-
-    /// An arbitrary pair of slices of a vector of length `dimension` _that have the same length_
-    fn arb_slice_pair(dimension: usize) -> impl Strategy<Value = [(usize, usize); 2]> {
-        // Similarly to `arb_slice`, we triplicate the entries because we want to also test the case
-        // where two or all three values coincide.
-        let all_indices: Vec<_> = (0..=dimension).flat_map(|i| [i, i, i]).collect();
-        proptest::sample::subsequence(all_indices, 3)
-            .prop_map(|v| [(v[0], v[1]), (v[2] - (v[1] - v[0]), v[2])])
-            .prop_shuffle()
-    }
-
-    /// An arbitrary vector of length `dimension` containing values in the range `0..p`
-    fn arb_vec_u32(p: ValidPrime, dimension: usize) -> impl Strategy<Value = Vec<u32>> {
-        proptest::collection::vec(0..*p, dimension)
+        (0..dimension).prop_flat_map(move |first| (Just(first), first..dimension))
```

Comment:
Does `move` here matter since `u32` implements `Copy`?

Reply:
I would have thought it was unnecessary but removing it makes the compiler believe that the closure outlives the function
```
error[E0373]: closure may outlive the current function, but it borrows `dimension`, which is owned by the current function
   --> crates/fp/src/vector/mod.rs:120:38
    |
120 |         (0..dimension).prop_flat_map(|first| (Just(first), first..dimension))
    |                                      ^^^^^^^                      --------- `dimension` is borrowed here
    |                                      |
    |                                      may outlive borrowed value `dimension`
    |
note: closure is returned here
   --> crates/fp/src/vector/mod.rs:120:9
    |
120 |         (0..dimension).prop_flat_map(|first| (Just(first), first..dimension))
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: to force the closure to take ownership of `dimension` (and any other referenced variables), use the `move` keyword
    |
120 |         (0..dimension).prop_flat_map(move |first| (Just(first), first..dimension))
    |                                      ++++
```

I think closures just always take values by reference, and we need to use `move` even when it's just a copy. That's what I understood from `rustc --explain E0373`

Reply:
Thanks for the explanation!

Reply:
No problem, I learned something too!